{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0857a73f",
   "metadata": {},
   "source": [
    "<center> Vizum HD data set compared between 2 to 8 micron "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a688d96d",
   "metadata": {},
   "source": [
    "## Analyzing Spatial Transcriptomics Data from Visium Output\n",
    "\n",
    "In this blog, we delve into the world of spatial transcriptomics, using the 'Mouse_Brain_square_example' provided by 10X Genomics. This example serves as a practical illustration of how spatial transcriptomic technologies can enhance our understanding of biological processes within specific tissue sections.\n",
    "\n",
    "## Comparison Between 2 Micron Bins and 8 Micron Bins\n",
    "\n",
    "An essential part of our analysis involves comparing data resolution at different scales. We will explore the distinctions between 2 micron bins and 8 micron bins. This comparison aims to highlight how the resolution affects our interpretation of the transcriptomic data, potentially impacting the conclusions we can draw about cellular functions and interactions within the tissue.\n",
    "\n",
    "## Displaying Different Representations of Areas with Low and High Read Counts\n",
    "\n",
    "Further, it is crucial to visualize how areas of low and high read counts are represented in the data. These visualizations help elucidate the distribution and density of transcriptomic activity across the tissue sample. By examining these differences, we can gain insights into the heterogeneity of gene expression and its implications for understanding tissue structure and function.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f8d7e08f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0564a65f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "|-- Visium_HD_Mouse_Brain_square_example\n",
      "  |-- ._.DS_Store\n",
      "  |-- ._square_008um\n",
      "  |-- ._square_002um\n",
      "  |-- square_008um\n",
      "    |-- ._analysis\n",
      "    |-- ._.DS_Store\n",
      "    |-- raw_feature_bc_matrix\n",
      "      |-- ._features.tsv.gz\n",
      "      |-- ._matrix.mtx.gz\n",
      "      |-- features.tsv.gz\n",
      "      |-- barcodes.tsv.gz\n",
      "      |-- ._barcodes.tsv.gz\n",
      "      |-- matrix.mtx.gz\n",
      "    |-- analysis\n",
      "      |-- clustering\n",
      "        |-- ._gene_expression_graphclust\n",
      "        |-- ._gene_expression_kmeans_4_clusters\n",
      "        |-- gene_expression_kmeans_6_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- ._gene_expression_kmeans_5_clusters\n",
      "        |-- ._gene_expression_kmeans_10_clusters\n",
      "        |-- gene_expression_kmeans_4_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- ._gene_expression_kmeans_2_clusters\n",
      "        |-- gene_expression_kmeans_2_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- gene_expression_kmeans_3_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- gene_expression_kmeans_10_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- ._gene_expression_kmeans_6_clusters\n",
      "        |-- gene_expression_kmeans_9_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- gene_expression_kmeans_7_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- ._gene_expression_kmeans_9_clusters\n",
      "        |-- gene_expression_kmeans_5_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- ._gene_expression_kmeans_8_clusters\n",
      "        |-- gene_expression_graphclust\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- ._gene_expression_kmeans_7_clusters\n",
      "        |-- gene_expression_kmeans_8_clusters\n",
      "          |-- clusters.csv\n",
      "          |-- ._clusters.csv\n",
      "        |-- ._gene_expression_kmeans_3_clusters\n",
      "      |-- ._diffexp\n",
      "      |-- ._umap\n",
      "      |-- pca\n",
      "        |-- ._gene_expression_10_components\n",
      "        |-- gene_expression_10_components\n",
      "          |-- components.csv\n",
      "          |-- variance.csv\n",
      "          |-- ._features_selected.csv\n",
      "          |-- ._variance.csv\n",
      "          |-- ._projection.csv\n",
      "          |-- projection.csv\n",
      "          |-- dispersion.csv\n",
      "          |-- ._components.csv\n",
      "          |-- ._dispersion.csv\n",
      "          |-- features_selected.csv\n",
      "      |-- ._pca\n",
      "      |-- umap\n",
      "        |-- ._gene_expression_2_components\n",
      "        |-- gene_expression_2_components\n",
      "          |-- ._projection.csv\n",
      "          |-- projection.csv\n",
      "      |-- ._clustering\n",
      "      |-- diffexp\n",
      "        |-- ._gene_expression_graphclust\n",
      "        |-- ._gene_expression_kmeans_4_clusters\n",
      "        |-- gene_expression_kmeans_6_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- ._gene_expression_kmeans_5_clusters\n",
      "        |-- ._gene_expression_kmeans_10_clusters\n",
      "        |-- gene_expression_kmeans_4_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- ._gene_expression_kmeans_2_clusters\n",
      "        |-- gene_expression_kmeans_2_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- gene_expression_kmeans_3_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- gene_expression_kmeans_10_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- ._gene_expression_kmeans_6_clusters\n",
      "        |-- gene_expression_kmeans_9_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- gene_expression_kmeans_7_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- ._gene_expression_kmeans_9_clusters\n",
      "        |-- gene_expression_kmeans_5_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- ._gene_expression_kmeans_8_clusters\n",
      "        |-- gene_expression_graphclust\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- ._gene_expression_kmeans_7_clusters\n",
      "        |-- gene_expression_kmeans_8_clusters\n",
      "          |-- ._differential_expression.csv\n",
      "          |-- differential_expression.csv\n",
      "        |-- ._gene_expression_kmeans_3_clusters\n",
      "    |-- filtered_feature_bc_matrix.h5\n",
      "    |-- ._raw_feature_bc_matrix\n",
      "    |-- ._filtered_feature_bc_matrix\n",
      "    |-- ._raw_feature_bc_matrix.h5\n",
      "    |-- cloupe.cloupe\n",
      "    |-- ._spatial\n",
      "    |-- .DS_Store\n",
      "    |-- filtered_feature_bc_matrix\n",
      "      |-- ._features.tsv.gz\n",
      "      |-- ._matrix.mtx.gz\n",
      "      |-- features.tsv.gz\n",
      "      |-- barcodes.tsv.gz\n",
      "      |-- ._barcodes.tsv.gz\n",
      "      |-- matrix.mtx.gz\n",
      "    |-- raw_feature_bc_matrix.h5\n",
      "    |-- spatial\n",
      "      |-- ._tissue_positions.parquet\n",
      "      |-- ._aligned_fiducials.jpg\n",
      "      |-- cytassist_image.tiff\n",
      "      |-- ._tissue_hires_image.png\n",
      "      |-- scalefactors_json.json\n",
      "      |-- tissue_hires_image.png\n",
      "      |-- ._detected_tissue_image.jpg\n",
      "      |-- ._scalefactors_json.json\n",
      "      |-- aligned_fiducials.jpg\n",
      "      |-- ._cytassist_image.tiff\n",
      "      |-- tissue_positions.parquet\n",
      "      |-- ._tissue_lowres_image.png\n",
      "      |-- aligned_tissue_image.jpg\n",
      "      |-- tissue_lowres_image.png\n",
      "      |-- detected_tissue_image.jpg\n",
      "      |-- ._aligned_tissue_image.jpg\n",
      "    |-- ._cloupe.cloupe\n",
      "    |-- ._filtered_feature_bc_matrix.h5\n",
      "  |-- .DS_Store\n",
      "  |-- square_002um\n",
      "    |-- raw_feature_bc_matrix\n",
      "      |-- ._features.tsv.gz\n",
      "      |-- ._matrix.mtx.gz\n",
      "      |-- features.tsv.gz\n",
      "      |-- barcodes.tsv.gz\n",
      "      |-- ._barcodes.tsv.gz\n",
      "      |-- matrix.mtx.gz\n",
      "    |-- filtered_feature_bc_matrix.h5\n",
      "    |-- ._raw_feature_bc_matrix\n",
      "    |-- ._filtered_feature_bc_matrix\n",
      "    |-- ._raw_feature_bc_matrix.h5\n",
      "    |-- ._spatial\n",
      "    |-- filtered_feature_bc_matrix\n",
      "      |-- ._features.tsv.gz\n",
      "      |-- ._matrix.mtx.gz\n",
      "      |-- features.tsv.gz\n",
      "      |-- barcodes.tsv.gz\n",
      "      |-- ._barcodes.tsv.gz\n",
      "      |-- matrix.mtx.gz\n",
      "    |-- raw_feature_bc_matrix.h5\n",
      "    |-- ._raw_probe_bc_matrix.h5\n",
      "    |-- raw_probe_bc_matrix.h5\n",
      "    |-- spatial\n",
      "      |-- ._tissue_positions.parquet\n",
      "      |-- ._aligned_fiducials.jpg\n",
      "      |-- cytassist_image.tiff\n",
      "      |-- ._tissue_hires_image.png\n",
      "      |-- scalefactors_json.json\n",
      "      |-- tissue_hires_image.png\n",
      "      |-- ._detected_tissue_image.jpg\n",
      "      |-- ._scalefactors_json.json\n",
      "      |-- aligned_fiducials.jpg\n",
      "      |-- ._cytassist_image.tiff\n",
      "      |-- tissue_positions.parquet\n",
      "      |-- ._tissue_lowres_image.png\n",
      "      |-- aligned_tissue_image.jpg\n",
      "      |-- tissue_lowres_image.png\n",
      "      |-- detected_tissue_image.jpg\n",
      "      |-- ._aligned_tissue_image.jpg\n",
      "    |-- ._filtered_feature_bc_matrix.h5\n"
     ]
    }
   ],
   "source": [
    "def display_directory_map(dir_path, indent=0):\n",
    "    # Print the current directory name\n",
    "    print('  ' * indent + '|-- ' + os.path.basename(dir_path))\n",
    "    # Get all items in the directory\n",
    "    for item in os.listdir(dir_path):\n",
    "        # Full path of the item\n",
    "        path = os.path.join(dir_path, item)\n",
    "        # Check if it's a directory and recurse into it\n",
    "        if os.path.isdir(path):\n",
    "            display_directory_map(path, indent + 1)\n",
    "        else:\n",
    "            # It's a file, print its name\n",
    "            print('  ' * (indent + 1) + '|-- ' + item)\n",
    "\n",
    "# Start directory map display from the specified path\n",
    "display_directory_map('/data/kanferg/Sptial_Omics/playGround/Data/Visium_HD_Mouse_Brain_square_example')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b875bb22",
   "metadata": {},
   "outputs": [],
   "source": [
    "path2mc = '/data/kanferg/Sptial_Omics/playGround/Data/Visium_HD_Mouse_Brain_square_example/square_002um'\n",
    "path8mc = '/data/kanferg/Sptial_Omics/playGround/Data/Visium_HD_Mouse_Brain_square_example/square_008um'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f7132fda",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "Could not find '/data/kanferg/Sptial_Omics/playGround/Data/Visium_HD_Mouse_Brain_square_example/square_008um/spatial/tissue_positions_list.csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1555179/1935280688.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0madata2mc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_visium\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpath8mc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0madata2mc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/scanpy/readwrite.py\u001b[0m in \u001b[0;36mread_visium\u001b[0;34m(path, genome, count_file, library_id, load_images, source_image_path)\u001b[0m\n\u001b[1;32m    451\u001b[0m                     )\n\u001b[1;32m    452\u001b[0m                 \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 453\u001b[0;31m                     \u001b[0;32mraise\u001b[0m \u001b[0mOSError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Could not find '{f}'\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    454\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    455\u001b[0m         \u001b[0madata\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muns\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"spatial\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mlibrary_id\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"images\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: Could not find '/data/kanferg/Sptial_Omics/playGround/Data/Visium_HD_Mouse_Brain_square_example/square_008um/spatial/tissue_positions_list.csv'"
     ]
    }
   ],
   "source": [
    "adata2mc = sc.read_visium(path = path8mc)\n",
    "adata2mc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c534f2f8",
   "metadata": {},
   "source": [
    "The missing file ```tissue_positions_list.csv``` is replaced with tissue_position.parquet. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c57dca98",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parquet_to_csv(path):\n",
    "    # Read the Parquet file\n",
    "    df = pd.read_parquet(os.path.join(path,'tissue_positions.parquet'))\n",
    "    # Write to a CSV file\n",
    "    df.to_csv(os.path.join(path,'tissue_positions_list.csv'), index=False)\n",
    "parquet_to_csv(path2mc + '/spatial')\n",
    "parquet_to_csv(path8mc + '/spatial')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fbf9b80d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/scanpy/readwrite.py:476: DtypeWarning: Columns (1,2,3,4,5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  positions = pd.read_csv(\n",
      "/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/gpfs/gsfs10/users/kanferg/conda/envs/squidpy/lib/python3.10/site-packages/scanpy/readwrite.py:476: DtypeWarning: Columns (1,2,3,4,5) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  positions = pd.read_csv(\n"
     ]
    }
   ],
   "source": [
    "adata2mc = sc.read_visium(path = path2mc)\n",
    "adata8mc = sc.read_visium(path = path8mc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9416c979",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6296688 × 19059\n",
       "    obs: 'in_tissue', 'array_row', 'array_col'\n",
       "    var: 'gene_ids', 'feature_types', 'genome'\n",
       "    uns: 'spatial'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2mc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7e47111",
   "metadata": {},
   "source": [
    "The ```adata``` shows that it found 6_296_688 n_obs (or cells or tiles)  19_059 n_vars (genes)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "28df9525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[9295.414883943817, 18095.4075623532],\n",
       "        [5556.6583091044295, 5393.048694317109],\n",
       "        [13608.742910510004, 2385.629487758882],\n",
       "        ...,\n",
       "        [5004.368167173434, 3620.2772970198544],\n",
       "        [13680.331660278203, 2289.6081731694144],\n",
       "        [9735.187714137932, 6533.6050688085215]], dtype=object),\n",
       " (6296688, 2))"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2mc.obsm['spatial'],adata2mc.obsm['spatial'].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fabdd1b2",
   "metadata": {},
   "source": [
    "```adata.obsm[‘spatial’]``` each coordinate corresponds to a specific cell for each feature. 6_296_688 by 2. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4b5a1d80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 6296688 × 19059\n",
       "    obs: 'in_tissue', 'array_row', 'array_col'\n",
       "    var: 'gene_ids', 'feature_types', 'genome'\n",
       "    uns: 'spatial'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2mc.var_names_make_unique()\n",
    "adata8mc.var_names_make_unique()\n",
    "adata2mc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fd177aa0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('spatial',\n",
       "              {'Visium_HD_Mouse_Brain': {'images': {'hires': array([[[0.92941177, 0.92941177, 0.92941177],\n",
       "                         [0.9254902 , 0.9254902 , 0.92941177],\n",
       "                         [0.92156863, 0.9254902 , 0.93333334],\n",
       "                         ...,\n",
       "                         [0.91764706, 0.91764706, 0.9254902 ],\n",
       "                         [0.9137255 , 0.91764706, 0.9254902 ],\n",
       "                         [0.9137255 , 0.91764706, 0.9254902 ]],\n",
       "                 \n",
       "                        [[0.9254902 , 0.9254902 , 0.93333334],\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.92156863, 0.9254902 , 0.93333334],\n",
       "                         [0.91764706, 0.92156863, 0.92941177],\n",
       "                         [0.91764706, 0.92156863, 0.92941177]],\n",
       "                 \n",
       "                        [[0.92941177, 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334],\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.92156863, 0.92156863, 0.92941177]],\n",
       "                 \n",
       "                        ...,\n",
       "                 \n",
       "                        [[0.92941177, 0.92941177, 0.92941177],\n",
       "                         [0.92156863, 0.9254902 , 0.92941177],\n",
       "                         [0.9254902 , 0.93333334, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.9254902 , 0.91764706, 0.9372549 ],\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.9254902 , 0.93333334, 0.92941177]],\n",
       "                 \n",
       "                        [[0.92941177, 0.92941177, 0.93333334],\n",
       "                         [0.92941177, 0.92156863, 0.9411765 ],\n",
       "                         [0.9254902 , 0.92156863, 0.9411765 ],\n",
       "                         ...,\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334],\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334]],\n",
       "                 \n",
       "                        [[0.92941177, 0.92941177, 0.9372549 ],\n",
       "                         [0.93333334, 0.9254902 , 0.93333334],\n",
       "                         [0.92941177, 0.9254902 , 0.9411765 ],\n",
       "                         ...,\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334]]], dtype=float32),\n",
       "                 'lowres': array([[[0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.92156863, 0.92156863, 0.92941177]],\n",
       "                 \n",
       "                        [[0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.92156863, 0.92156863, 0.92941177]],\n",
       "                 \n",
       "                        [[0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.92156863, 0.92156863, 0.92941177],\n",
       "                         [0.92156863, 0.92156863, 0.92941177]],\n",
       "                 \n",
       "                        ...,\n",
       "                 \n",
       "                        [[0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.92941177, 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334],\n",
       "                         [0.9254902 , 0.9254902 , 0.92941177],\n",
       "                         [0.92156863, 0.9254902 , 0.92941177]],\n",
       "                 \n",
       "                        [[0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.92941177, 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334],\n",
       "                         [0.92156863, 0.9254902 , 0.93333334],\n",
       "                         [0.92156863, 0.9254902 , 0.93333334]],\n",
       "                 \n",
       "                        [[0.9254902 , 0.92941177, 0.93333334],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         [0.9254902 , 0.92941177, 0.9372549 ],\n",
       "                         ...,\n",
       "                         [0.9254902 , 0.9254902 , 0.93333334],\n",
       "                         [0.92156863, 0.9254902 , 0.93333334],\n",
       "                         [0.92156863, 0.9254902 , 0.92941177]]], dtype=float32)},\n",
       "                'scalefactors': {'spot_diameter_fullres': 7.303953797779634,\n",
       "                 'bin_size_um': 2.0,\n",
       "                 'microns_per_pixel': 0.2738242950835738,\n",
       "                 'regist_target_img_scalef': 0.2505533,\n",
       "                 'tissue_lowres_scalef': 0.02505533,\n",
       "                 'fiducial_diameter_fullres': 1205.1523766336395,\n",
       "                 'tissue_hires_scalef': 0.2505533},\n",
       "                'metadata': {'chemistry_description': 'Visium HD v1',\n",
       "                 'software_version': 'spaceranger-3.0.0'}}})])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata2mc.uns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26e5aecf",
   "metadata": {},
   "source": [
    "[Analysis and visualization of spatial transcriptomics data](https://scanpy.readthedocs.io/en/stable/tutorials/spatial/basic-analysis.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c71d7394",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata2mc.var[\"mt\"] = adata2mc.var_names.str.startswith(\"mt-\")\n",
    "adata8mc.var[\"mt\"] = adata8mc.var_names.str.startswith(\"mt-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "c794cd9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ids</th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>mt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>ENSMUSG00000025900</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sox17</th>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lypla1</th>\n",
       "      <td>ENSMUSG00000025903</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tcea1</th>\n",
       "      <td>ENSMUSG00000033813</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd4</th>\n",
       "      <td>ENSMUSG00000064363</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd5</th>\n",
       "      <td>ENSMUSG00000064367</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Nd6</th>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mt-Cytb</th>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vamp7</th>\n",
       "      <td>ENSMUSG00000051412</td>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>mm10</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19059 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_ids    feature_types genome     mt\n",
       "Xkr4     ENSMUSG00000051951  Gene Expression   mm10  False\n",
       "Rp1      ENSMUSG00000025900  Gene Expression   mm10  False\n",
       "Sox17    ENSMUSG00000025902  Gene Expression   mm10  False\n",
       "Lypla1   ENSMUSG00000025903  Gene Expression   mm10  False\n",
       "Tcea1    ENSMUSG00000033813  Gene Expression   mm10  False\n",
       "...                     ...              ...    ...    ...\n",
       "mt-Nd4   ENSMUSG00000064363  Gene Expression   mm10   True\n",
       "mt-Nd5   ENSMUSG00000064367  Gene Expression   mm10   True\n",
       "mt-Nd6   ENSMUSG00000064368  Gene Expression   mm10   True\n",
       "mt-Cytb  ENSMUSG00000064370  Gene Expression   mm10   True\n",
       "Vamp7    ENSMUSG00000051412  Gene Expression   mm10  False\n",
       "\n",
       "[19059 rows x 4 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2822b239",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
